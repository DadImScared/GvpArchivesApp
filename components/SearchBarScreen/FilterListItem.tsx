import * as React from "react";
import { Platform, StyleSheet, View } from "react-native";

import { CheckBox, ListItem, NativeBase, Text } from "native-base";

import { SearchStateAndActions } from "../../reducers/search";
import { routes } from "../SideBar";

const styles = StyleSheet.create({
  innerListWrapper: {
    alignContent: "flex-start",
    alignItems: "flex-start",
    alignSelf: "flex-start",
    flexDirection: "row",
    justifyContent: "center",
    paddingRight: 0
  },
  text: {
    alignSelf: "center",
    marginLeft: 8,
    marginRight: 0,
    minWidth: 100
  }
});

interface IProps extends NativeBase.ListItem, SearchStateAndActions<"toggleCategory"> {
  category: string;
  addRow: (category: string, width: number) => void;
}

export class FilterListItem extends React.Component<IProps> {

  toggleCategory = Platform.select<any>({
    default: () => this.props.toggleCategory(this.props.category),
    ios: () => setTimeout(() => this.props.toggleCategory(this.props.category), 50)
  });

  addToRow = ({ nativeEvent: { layout: { width } } }: any) => {
    this.props.addRow(this.props.category, width + 36);
  };

  render() {
    const { category, toggleCategory, addRow, ...other } = this.props;
    return (
      <ListItem button={true} onPress={this.toggleCategory} {...other}>
        <View style={styles.innerListWrapper} onLayout={this.addToRow}>
          <CheckBox checked={other.selected} color={routes[category].icon.style.color} />
          <Text style={styles.text}>{routes[category].displayName}</Text>
        </View>
      </ListItem>
    );
  }
}

export default FilterListItem;
