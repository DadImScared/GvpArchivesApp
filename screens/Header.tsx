import * as React from "react";

import { Ionicons } from "@expo/vector-icons";
import { NavigationScreenProps } from "react-navigation";
import HeaderButtons, { HeaderButton } from "react-navigation-header-buttons";

const IoniconsHeaderButton = (props: any) => (
  <HeaderButton {...props} IconComponent={Ionicons} iconSize={23} color={"blue"} />
);

export const IoniconsHeaderButtons = (props: any) => (
  <HeaderButtons
    HeaderButtonComponent={IoniconsHeaderButton}
    {...props}
  />
);

export const HeaderItem = HeaderButtons.Item;

class HeaderRight extends React.Component<NavigationScreenProps> {

  searchPress = () => {
    this.props.navigation.navigate("SearchBar");
  };

  render() {
    return (
      <IoniconsHeaderButtons>
        <HeaderItem iconName={"ios-search"} onPress={this.searchPress} title={"Search"} />
      </IoniconsHeaderButtons>
    );
  }
}

export const headerRight = ({ navigation }: NavigationScreenProps) => (
  <HeaderRight navigation={navigation} />
);
