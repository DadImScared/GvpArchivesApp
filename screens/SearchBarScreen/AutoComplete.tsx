import * as React from "react";
import { FlatList, ListRenderItem, Text, View } from "react-native";
import { NavigationScreenProps } from "react-navigation";
import { connect } from "react-redux";
import { compose } from "recompose";

import { Spinner } from "native-base";

import { IReducerState } from "../../reducers";
import { getAutoCompleteScreen, IAutoCompleteScreenData } from "../../reducers/autoComplete";

import ListItem from "../../components/SearchBarScreen/AutoCompleteListItem";

class AutoComplete extends React.Component<IAutoCompleteScreenData & NavigationScreenProps> {
  renderListItem: ListRenderItem<string> = ({ item }) => (
    <ListItem text={item} />
  );

  keyExtractor = (item: string, index: number) => `${item}-${this.props.queryId}-${index}`;

  render() {
    const { isLoading, query, autoComplete: { results } } = this.props;
    return (
      <View>
        {
          query ?
            isLoading || isLoading === undefined ?
              <Spinner color={"green"}/>
              :
              results.length ?
                <FlatList
                  keyExtractor={this.keyExtractor}
                  data={results}
                  renderItem={this.renderListItem}
                />
                :
                <Text>No suggestions</Text>
            :
            <Text>Type in the search bar to see suggestions!</Text>
        }
      </View>
    );
  }
}

const mapStateToProps = (state: IReducerState) => getAutoCompleteScreen(state);

const enhance = compose<any, any>(
  connect(mapStateToProps)
);

export default enhance(AutoComplete);
