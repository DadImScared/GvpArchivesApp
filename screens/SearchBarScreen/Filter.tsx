import * as React from "react";
import { Text, View } from "react-native";

export class Filter extends React.Component {
  render() {
    return (
      <View>
        <Text>filter</Text>
      </View>
    );
  }
}

export default Filter;
