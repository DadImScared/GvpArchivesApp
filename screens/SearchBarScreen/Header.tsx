import * as React from "react";
import { BackHandler, StyleSheet, View } from "react-native";
import { NavigationScreenProps } from "react-navigation";
import { connect } from "react-redux";
import { Dispatch } from "redux";

import _ from "lodash";
import { Header as BaseHeader, Input, Item } from "native-base";

import Icon from "../../components/Icon";

import { search } from "../../actions";
import { getQuerySuggestions } from "../../actions/search";
import { IReducerState } from "../../reducers";
import { getQueryId, getSearch, SearchStateAndActions } from "../../reducers/search";

interface IProps extends NavigationScreenProps, SearchStateAndActions<"query" | "updateSearchQuery" | "categories"> {
  getQuerySuggestions: (query: string, categories: string[]) => any;
}

const styles: any = StyleSheet.create({
  header: {
    backgroundColor: "white",
    borderBottomWidth: 0,
    height: 40,
    paddingTop: 0
  }
});

export class Header extends React.Component<IProps> {
  textRef!: any;
  didFocus!: any;
  willBlur!: any;

  querySuggestions = _.debounce(this.props.getQuerySuggestions, 150);

  componentDidMount() {
    const { query, categories } = this.props;
    this.didFocus = this.props.navigation.addListener("didFocus", () => {
      this.textRef && this.textRef._root.focus();
      BackHandler.addEventListener("hardwareBackPress", this.handleBackButton);
    });
    this.willBlur = this.props.navigation.addListener("willBlur", () => {
      // console.log("will blur");
      BackHandler.removeEventListener("hardwareBackPress", this.handleBackButton);
    });

    if (query) {
      this.props.getQuerySuggestions(query, categories);
    }
  }

  componentDidUpdate(prevProps: Readonly<IProps>, prevState: Readonly<{}>, snapshot?: any): void {
    const { navigation: { state: { routes } }, query, categories } = this.props;
    const { index } = routes[0];
    // when auto complete screen re-focuses fetch data
    if ((index !== prevProps.navigation.state.routes[0].index) && index === 0) {
      this.props.getQuerySuggestions(query, categories);
    }
  }

  componentWillUnmount(): void {
    BackHandler.removeEventListener("hardwareBackPress", this.handleBackButton);
    this.didFocus && this.didFocus.remove();
    this.willBlur && this.willBlur.remove();
  }

  setTextRef = (el: Input) => this.textRef = el;

  handleBackButton = () => {
    this.props.navigation.goBack();
    return true;
  };

  updateQuery = (query: string) => {
    this.props.updateSearchQuery(query);
    this.querySuggestions(query, this.props.categories);
  };

  navigateToSearch = () => {
    const { categories, query } = this.props;
    const parentNavigation = this.props.navigation.dangerouslyGetParent();
    const params = { query, categories, queryId: getQueryId(query, categories) };
    if (parentNavigation) parentNavigation.replace("Search", params);
    // this.props.navigation.push("Search", { query, categories, queryId: getQueryId(query, categories) });
  };

  render() {
    return (
      <View style={{ width: "100%" }}>
        <BaseHeader noShadow={true} style={styles.header} searchBar={true}>
          <Item rounded={true}>
            <Icon name={"ios-search"} />
            <Input
              value={this.props.query}
              onSubmitEditing={this.navigateToSearch}
              onChangeText={this.updateQuery}
              ref={this.setTextRef}
              placeholder={"Search"}
            />
          </Item>
        </BaseHeader>
      </View>
    );
  }
}

const mapStateToProps = (state: IReducerState) => getSearch(state);

const mapDispatchToProps = (dispatch: Dispatch) => ({
  getQuerySuggestions: (...args: any[]) => dispatch((getQuerySuggestions as any)(...args)),
  updateSearchQuery: (query: string) => dispatch(search.updateSearchQuery(query))
});

export default connect(mapStateToProps, mapDispatchToProps)(Header);
