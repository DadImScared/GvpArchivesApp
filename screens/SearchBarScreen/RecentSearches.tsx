import * as React from "react";
import { Text, View } from "react-native";

export class RecentSearches extends React.Component {
  render() {
    return (
      <View>
        <Text>Recent searches</Text>
      </View>
    );
  }
}

export default RecentSearches;
