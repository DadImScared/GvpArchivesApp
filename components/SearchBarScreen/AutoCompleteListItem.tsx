import * as React from "react";
import { Text } from "react-native";
import { NavigationScreenProp } from "react-navigation";

import { ListItem } from "native-base";

import { ISearchState } from "../../reducers/search";
import { searchParams } from "../../screens/Search";

interface IProps extends ISearchState {
  text: string;
  navigation: NavigationScreenProp<any>;
}

export class AutoCompleteListItem extends React.Component<IProps> {

  navigateSearch = () => {
    const { navigation, text, categories } = this.props;
    const parentNavigation = (navigation as any).dangerouslyGetParent().dangerouslyGetParent().dangerouslyGetParent();
    parentNavigation.replace("Search", searchParams(text, categories));
  };

  render() {
    return (
      <ListItem button={true} noIndent={true} onPress={this.navigateSearch}>
        <Text>{this.props.text}</Text>
      </ListItem>
    );
  }
}

export default AutoCompleteListItem;
