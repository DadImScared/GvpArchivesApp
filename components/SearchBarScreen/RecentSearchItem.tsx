import * as React from "react";
import { Platform, View } from "react-native";
import { NavigationScreenProp } from "react-navigation";

import { Body, H2, ListItem, Text } from "native-base";

import { IRecentSearch } from "../../reducers/recentSearches";
import { searchParams } from "../../screens/Search";
import { wrapper } from "../../styles";
import Icon from "../Icon";
import { routes } from "../SideBar";

interface IProps {
  search: IRecentSearch;
  navigation: NavigationScreenProp<any>;
}

export class RecentSearchItem extends React.Component<IProps> {

  navigateSearch = () => {
    const { search, navigation } = this.props;
    const parentNavigation = (navigation as any).dangerouslyGetParent().dangerouslyGetParent().dangerouslyGetParent();
    parentNavigation.replace("Search", searchParams(search.query, search.categories));
  };

  render() {
    const { search: { categories, query, timestamp } } = this.props;
    return (
      <ListItem button={true} onPress={this.navigateSearch}>
        <Body>
          <H2>{query}</H2>
          <View style={wrapper.inline}>
            {
              categories.map((category, index) => (
                <Text
                  key={`${query}-${timestamp}-${index}`}
                  style={{
                    backgroundColor: routes[category].icon.style.color,
                    color: "white",
                    fontSize: 14,
                    marginBottom: 0,
                    marginLeft: 0,
                    marginRight: 4,
                    marginTop: 4,
                    padding: 4,
                    ...Platform.select<any>({
                      android: {
                        borderRadius: 25
                      },
                      ios: {
                        borderRadius: 12,
                        overflow: "hidden"
                      }
                    }),
                  }}
                >
                  {routes[category].displayName.toUpperCase()}
                </Text>
              ))
            }
          </View>
        </Body>
        <Icon name={"arrow-forward"} />
      </ListItem>
    );
  }
}

export default RecentSearchItem;
