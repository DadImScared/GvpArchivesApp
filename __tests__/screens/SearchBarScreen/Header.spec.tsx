import * as React from "react";
import * as renderer from "react-test-renderer";

jest.useFakeTimers();

import { Input } from "native-base";
import { Header } from "../../../screens/SearchBarScreen/Header";

describe("Header", () => {
  let props: any;

  beforeEach(() => {
    const navActions = {
      addListener: jest.fn(),
      dangerouslyGetParent: () => navActions,
      replace: jest.fn()
    };
    props = {
      categories: ["song", "book"],
      getQuerySuggestions: jest.fn(),
      navigation: navActions,
      query: "radha",
      updateSearchQuery: jest.fn()
    };
  });

  it("should get suggestions on mount if query exists", () => {
    renderer.create(<Header {...props} />).toJSON();
    expect(props.getQuerySuggestions).toHaveBeenCalledTimes(1);
    expect(props.getQuerySuggestions).toBeCalledWith(props.query, props.categories);
  });

  it("should update search query on input", () => {
    const tree = renderer.create(<Header {...props} />);
    const testInstance = tree.root;
    const instanceProps = testInstance.findByType(Input).props;
    instanceProps.onChangeText("new query");
    expect(props.updateSearchQuery).toHaveBeenCalledTimes(1);
    expect(props.updateSearchQuery).toBeCalledWith("new query");
    jest.runAllTimers();
    expect(props.getQuerySuggestions).toHaveBeenCalledTimes(2);
    expect(props.getQuerySuggestions).toHaveBeenCalledWith("new query", props.categories);
  });

  it("should navigate on submit editing", () => {
    const tree = renderer.create(<Header {...props} />);
    const testInstance = tree.root;
    const instanceProps = testInstance.findByType(Input).props;
    instanceProps.onSubmitEditing();
    expect(props.navigation.replace.mock.calls[0]).toMatchSnapshot();
  });
});
