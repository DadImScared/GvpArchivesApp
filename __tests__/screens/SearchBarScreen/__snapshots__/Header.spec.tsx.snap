// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should navigate on submit editing 1`] = `
Array [
  "Search",
  Object {
    "categories": Array [
      "song",
      "book",
    ],
    "query": "radha",
    "queryId": "radha-book,song",
  },
]
`;
